monitorcontroller/monitorcontroller.go

```diff
@@ -98,16 +105,74 @@ func (monitor *MonitorController) Close() {
 	}
 }
 
-// SendMonitoringData sends monitoring data.
-func (monitor *MonitorController) SendMonitoringData(monitoringData cloudprotocol.NodeMonitoringData) {
+// SendNodeMonitoring sends monitoring data.
+func (monitor *MonitorController) SendNodeMonitoring(nodeMonitoring aostypes.NodeMonitoring) {
 	monitor.Lock()
-	defer monitor.Unlock()
 
-	if len(monitor.monitoringQueue) >= monitor.monitoringQueueSize {
-		monitor.monitoringQueue = monitor.monitoringQueue[1:]
+	// calculate size of input parameter
+	messageSize := 0
+
+	message, err := json.Marshal(nodeMonitoring)
+	if err == nil {
+		messageSize = len(message)
+	} else {
+		log.Errorf("Can't marshal nodeMonitoring: %v", err)
+	}
+
+	// allocate new offline message
+	if len(monitor.offlineMessages) == 0 || monitor.currentMessageSize+messageSize > monitor.maxMessageSize {
+		if len(monitor.offlineMessages) == cap(monitor.offlineMessages) {
+			monitor.offlineMessages = monitor.offlineMessages[1:]
+		}
+
+		monitor.offlineMessages = append(monitor.offlineMessages, cloudprotocol.Monitoring{})
+		monitor.currentMessageSize = 0
+	}
+
+	monitor.currentMessageSize += messageSize
+	latestMessage := &monitor.offlineMessages[len(monitor.offlineMessages)-1]
+
+	// add node monitoring data
+	nodeDataInd := slices.IndexFunc(latestMessage.Nodes,
+		func(item cloudprotocol.NodeMonitoringData) bool { return item.NodeID == nodeMonitoring.NodeID },
+	)
+
+	if nodeDataInd >= 0 {
+		nodeItems := &latestMessage.Nodes[nodeDataInd].Items
+		*nodeItems = append(*nodeItems, nodeMonitoring.NodeData)
+	} else {
+		latestMessage.Nodes = append(latestMessage.Nodes,
+			cloudprotocol.NodeMonitoringData{
+				NodeID: nodeMonitoring.NodeID, Items: []aostypes.MonitoringData{nodeMonitoring.NodeData},
+			},
+		)
+	}
+
+	// add instance monitoring data
+	for _, instanceData := range nodeMonitoring.InstancesData {
+		instanceDataInd := slices.IndexFunc(latestMessage.ServiceInstances,
+			func(item cloudprotocol.InstanceMonitoringData) bool {
+				return item.NodeID == nodeMonitoring.NodeID && item.InstanceIdent == instanceData.InstanceIdent
+			},
+		)
+
+		if instanceDataInd >= 0 {
+			serviceInstanceItems := &latestMessage.ServiceInstances[instanceDataInd].Items
+			*serviceInstanceItems = append(*serviceInstanceItems, instanceData.MonitoringData)
+		} else {
+			latestMessage.ServiceInstances = append(latestMessage.ServiceInstances,
+				cloudprotocol.InstanceMonitoringData{
+					NodeID:        nodeMonitoring.NodeID,
+					InstanceIdent: instanceData.InstanceIdent,
+					Items:         []aostypes.MonitoringData{instanceData.MonitoringData},
+				},
+			)
+		}
 	}
 
-	monitor.monitoringQueue = append(monitor.monitoringQueue, monitoringData)
+	// send notification message
+	monitor.Unlock()
+	monitor.sendMessageEvent <- struct{}{}
```

Comment:
Looks like in this implementation monitoring will be sent on each incoming node monitoring. Won't be?

Reply:
yes, according to:
[Jul 13, 2024](https://github.com/aosedge/aos_communicationmanager/pull/134#discussion_r1676840690)

We should send immediately and do not wait when offline queue is full

Reply:
changed to send event on current message overflow